File: walkgen_footstep_planner/python/walkgen_footstep_planner/params.py
#!/usr/bin/env python3
#
# Copyright 2022 University of Edinburgh
# All rights reserved.
#
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions are met:
#
#  * Redistributions of source code must retain the above copyright notice,
#    this list of conditions and the following disclaimer.
#  * Redistributions in binary form must reproduce the above copyright
#    notice, this list of conditions and the following disclaimer in the
#    documentation and/or other materials provided with the distribution.
#  * Neither the name of  nor the names of its contributors may be used to
#    endorse or promote products derived from this software without specific
#    prior written permission.
#
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
# ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
# LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
# CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
# SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
# POSSIBILITY OF SUCH DAMAGE.

import yaml


class FootStepPlannerParams:

    def __init__(self, filename=None):
        """ Parameters for the reactive footstep planning.

        Args:
            - filename (string): path to the config file.
        """

        # Number of phases returned by the SurfacePlanner
        # (N_phase_return surfaces for each foot)
        self.N_phase_return = 3

        # Gait parameters
        self.typeGait = "walk"  # Only "walk" or "trot" working
        self.dt = 0.01
        self.N_ss = 90
        self.N_ds = 70
        self.N_uds = 0
        self.N_uss = 0
        # (int or None), use the lenght of the gait (None) or a defined horizon (int)
        self.horizon = None
        self.nsteps = 1  # Number of iteration.
        self.stepHeight = 0.15  # Step height [m]

        # Bezier parameters
        self.margin_down = 0.10  # Margin [m] wrt to the segment crossed in the surface.
        self.t_margin_down = 0.2  # % of the curve constrained around critical point.
        self.z_margin_down = 0.2  # % of the curve after the critical point.
        self.margin_up = 0.10  # Margin [m] wrt to the segment crossed in the surface.
        self.t_margin_up = 0.045  # % of the curve constrained around critical point.
        self.z_margin_up = 0.045  # % of the curve after the critical point.

        self.N_sample = 10  # Number of sample in the least square optimisation for Bezier coeffs.
        self.N_sample_ineq = 8  # Number of sample while browsing the curve.
        self.degree = 7  # Degree of the Bezier curve.

        if filename is not None:
            self.parseFile(filename)

    def parseFile(self, filename):
        """ Parse yaml config file to setup the parameters.

        Args:
            - filename (string): path to the config file.
        """
        config = yaml.load(open(filename, 'r'), Loader=yaml.FullLoader)
        self.N_phase_return = config["walkgen_params"]["params"]["N_phase_return"]
        self.typeGait = config["walkgen_params"]["gait"]["type"]
        self.dt = config["walkgen_params"]["gait"]["dt"]
        self.N_ds = config["walkgen_params"]["gait"][self.typeGait]["N_ds"]
        self.N_ss = config["walkgen_params"]["gait"][self.typeGait]["N_ss"]
        self.N_uds = config["walkgen_params"]["gait"][self.typeGait]["N_uds"]
        self.N_uss = config["walkgen_params"]["gait"][self.typeGait]["N_uss"]
        self.horizon = config["walkgen_params"]["gait"]["horizon"]
        self.nsteps = config["walkgen_params"]["gait"]["nsteps"]
        self.stepHeight = config["walkgen_params"]["gait"]["stepHeight"]
        self.margin = config["walkgen_params"]["bezier"]["margin"]
        self.t_margin = config["walkgen_params"]["bezier"]["t_margin"]
        self.z_margin = config["walkgen_params"]["bezier"]["z_margin"]
        self.N_sample = config["walkgen_params"]["bezier"]["N_sample"]
        self.N_sample_ineq = config["walkgen_params"]["bezier"]["N_sample_ineq"]
        self.degree = config["walkgen_params"]["bezier"]["degree"]

File: walkgen_footstep_planner/src/Params.cpp
#include "Params.hpp"

using namespace yaml_check;

void Params::initialize_default() {
  type = "trot";
  dt = 0.01;
  horizon = 125;
  nsteps = 1;
  stepHeight = 0.15;
  N_ds = 80;
  N_ss = 35;
  N_uds = 0;
  N_uss = 0;
  N_phase_return = 2;
  margin_up = 0.1;
  t_margin_up = 0.045;
  z_margin_up = 0.045;
  margin_down = 0.1;
  t_margin_down = 0.2;
  z_margin_down = 0.2;
  N_sample = 10;
  N_sample_ineq = 8;
  degree = 7;
  feet_names = {"LF_FOOT", "LH_FOOT", "RF_FOOT", "RH_FOOT"};
  feet_names_sl1m = {"LF_FOOT", "LH_FOOT", "RF_FOOT", "RH_FOOT"};
  shoulder_offsets.push_back({0.367, 0.2});
  shoulder_offsets.push_back({0.367, -0.2});
  shoulder_offsets.push_back({-0.367, 0.2});
  shoulder_offsets.push_back({-0.367, -0.2});

  walk_N_ds = 90;
  walk_N_ss = 70;
  walk_N_uds = 0;
  walk_N_uss = 0;
  trot_N_ds = 40;
  trot_N_ss = 30;
  trot_N_uds = 0;
  trot_N_uss = 0;
}

Params::~Params() {}

Params::Params() { initialize_default(); }

Params::Params(const std::string &filename) {
  initialize_default();
  parse_yaml_file(filename);
}

Params::Params(const Params &other) {
  // Copy all member variables from other to this.
  this->type = other.type;
  this->dt = other.dt;
  this->horizon = other.horizon;
  this->nsteps = other.nsteps;
  this->stepHeight = other.stepHeight;
  this->N_ds = other.N_ds;
  this->N_ss = other.N_ss;
  this->N_uds = other.N_uds;
  this->N_uss = other.N_uss;
  this->N_phase_return = other.N_phase_return;
  this->feet_names = other.feet_names;
  this->feet_names_sl1m = other.feet_names_sl1m;
  this->shoulder_offsets = other.shoulder_offsets;
  this->margin_up = other.margin_up;
  this->t_margin_up = other.t_margin_up;
  this->z_margin_up = other.z_margin_up;
  this->margin_down = other.margin_down;
  this->t_margin_down = other.t_margin_down;
  this->z_margin_down = other.z_margin_down;
  this->N_sample = other.N_sample;
  this->N_sample_ineq = other.N_sample_ineq;
  this->degree = other.degree;

  // Gait parametres for changing
  this->walk_N_ds = other.walk_N_ds;
  this->walk_N_ss = other.walk_N_ss;
  this->walk_N_uds = other.walk_N_uds;
  this->walk_N_uss = other.walk_N_uss;
  this->trot_N_ds = other.trot_N_ds;
  this->trot_N_ss = other.trot_N_ss;
  this->trot_N_uds = other.trot_N_uds;
  this->trot_N_uss = other.trot_N_uss;
}

void Params::parse_yaml_file(const std::string &filename) {
  assert_file_exists(filename);
  YAML::Node yaml_node = YAML::LoadFile(filename);

  assert_yaml_parsing(yaml_node, filename, "walkgen_params");
  const YAML::Node &config = yaml_node["walkgen_params"];

  assert_yaml_parsing(config["params"], filename, "N_phase_return");
  N_phase_return = config["params"]["N_phase_return"].as<int>();

  // assert_yaml_parsing(config, filename, "gait");
  assert_yaml_parsing(config["gait"], filename, "type");
  assert_yaml_parsing(config["gait"], filename, "dt");
  assert_yaml_parsing(config["gait"], filename, "horizon");
  assert_yaml_parsing(config["gait"], filename, "nsteps");
  assert_yaml_parsing(config["gait"], filename, "stepHeight");
  assert_yaml_parsing(config["gait"], filename, "feet_names");
  assert_yaml_parsing(config["gait"], filename, "feet_names_sl1m");
  assert_yaml_parsing(config["gait"], filename, "shoulder_offsets");
  type = config["gait"]["type"].as<std::string>();
  dt = config["gait"]["dt"].as<double>();
  horizon = config["gait"]["horizon"].as<int>();
  nsteps = config["gait"]["nsteps"].as<int>();
  stepHeight = config["gait"]["stepHeight"].as<double>();
  feet_names = config["gait"]["feet_names"].as<std::vector<std::string>>();
  feet_names_sl1m = config["gait"]["feet_names_sl1m"].as<std::vector<std::string>>();
  shoulder_offsets = config["gait"]["shoulder_offsets"].as<std::vector<std::vector<double>>>();
  if (feet_names.size() != shoulder_offsets.size()) {
    throw std::runtime_error(
        "The contact name list and the shoulder offsets should "
        "have the same size.");
  }
  if (feet_names.size() != feet_names_sl1m.size()) {
    throw std::runtime_error("Both nominal contact list and sl1m should have the same size.");
  }
  for (auto elem : shoulder_offsets) {
    if (elem.size() != 2) {
      throw std::runtime_error(
          "The size of each offset in shoulder offset params "
          "should be size 2 (x and y axis).");
    }
  }
  if (type == "trot") {
    // Check the yaml file
    assert_yaml_parsing(config["gait"]["trot"], filename, "N_ds");
    assert_yaml_parsing(config["gait"]["trot"], filename, "N_ss");
    assert_yaml_parsing(config["gait"]["trot"], filename, "N_uds");
    assert_yaml_parsing(config["gait"]["trot"], filename, "N_uss");

    N_ds = config["gait"]["trot"]["N_ds"].as<int>();
    N_ss = config["gait"]["trot"]["N_ss"].as<int>();
    N_uds = config["gait"]["trot"]["N_uds"].as<int>();
    N_uss = config["gait"]["trot"]["N_uss"].as<int>();
  } else if (type == "walk") {
    // Check the yaml file
    assert_yaml_parsing(config["gait"]["walk"], filename, "N_ds");
    assert_yaml_parsing(config["gait"]["walk"], filename, "N_ss");
    assert_yaml_parsing(config["gait"]["walk"], filename, "N_uds");
    assert_yaml_parsing(config["gait"]["walk"], filename, "N_uss");

    N_ds = config["gait"]["walk"]["N_ds"].as<int>();
    N_ss = config["gait"]["walk"]["N_ss"].as<int>();
    N_uds = config["gait"]["walk"]["N_uds"].as<int>();
    N_uss = config["gait"]["walk"]["N_uss"].as<int>();
  }

  // Check the yaml file
  assert_yaml_parsing(config["bezier"], filename, "margin_up");
  assert_yaml_parsing(config["bezier"], filename, "t_margin_up");
  assert_yaml_parsing(config["bezier"], filename, "z_margin_up");
  assert_yaml_parsing(config["bezier"], filename, "margin_down");
  assert_yaml_parsing(config["bezier"], filename, "t_margin_down");
  assert_yaml_parsing(config["bezier"], filename, "z_margin_down");
  assert_yaml_parsing(config["bezier"], filename, "N_sample_ineq");
  assert_yaml_parsing(config["bezier"], filename, "degree");

  margin_up = config["bezier"]["margin_up"].as<double>();
  t_margin_up = config["bezier"]["t_margin_up"].as<double>();
  z_margin_up = config["bezier"]["z_margin_up"].as<double>();
  margin_down = config["bezier"]["margin_down"].as<double>();
  t_margin_down = config["bezier"]["t_margin_down"].as<double>();
  z_margin_down = config["bezier"]["z_margin_down"].as<double>();
  N_sample = config["bezier"]["N_sample"].as<int>();
  N_sample_ineq = config["bezier"]["N_sample_ineq"].as<int>();
  degree = config["bezier"]["degree"].as<int>();
}
